core/src/main/scala/cats/syntax/apply.scala

```diff
@@ -40,6 +35,22 @@ private[syntax] trait ApplySyntaxBinCompat0 {
     new IfApplyOps[F](fa)
 }
 
+final class ApplySyntaxBinCompat1[F[_], A](private val fa: F[A]) extends AnyVal {
```

Comment:
I'm not 100% sure if this will work, but if it does, then we don't have to copy all the methods here.

```suggestion
final class ApplySyntaxBinCompat1[F[_], A](private val fa: F[A]) extends AnyVal with Apply.Ops[F, A] {
```

Reply:
hmm, the compiler seems to complain about this change.

```
[error] /Users/s15255/work/cats/core/src/main/scala/cats/syntax/apply.scala:46:86: illegal inheritance; superclass AnyVal
[error] is not a subclass of the superclass Object
[error] of the mixin trait Ops
[error] final class ApplyBinCompat1[F[_], A](private val fa: F[A]) extends AnyVal with Apply.Ops[F, A] {
[error] ^
[error] one error found
[error] (coreJVM / Compile / compileIncremental) Compilation failed
```

maybe because `Apply.Ops` is not a [universal trait](https://docs.scala-lang.org/overviews/core/value-classes.html)?

> A value class may not extend a non-universal trait

Reply:
yes, `Apply.Ops` has to be a universal trait to be allowed mixing with `AnyVal`.

Reply:
Oh interesting. Can we change that compatibly? Maybe not 🤔

Reply:
I doubt it. It has a `val` defined inside whereas afaik universal traits are not allowed to contain vals:
```
  trait Ops[F[_], A] extends Serializable {
    type TypeClassType <: Apply[F]
    def self: F[A]
    val typeClassInstance: TypeClassType
    ...
```

Reply:
... however... in theory we can try the opposite – define our own universal trait with all the methods we need and inherit both the legacy `Apply.Ops` and the new `ApplySyntax` from it... it may or may not help though – I'm not sure it will.

Reply:
> define our own universal trait with all the methods we need and inherit both the legacy Apply.Ops and the new ApplySyntax from it

sounds good, I'll try that.

Reply:
I tried this a bit, but there seems to be another problem when making old `Ops` inherit universalaltrait :(
I’m not sure, but maybe mima complaining about [this case](https://github.com/lightbend/mima/tree/main/functional-tests/src/test/trait-deleting-concrete-methods-is-nok), which implies we can’t remove old Ops’s concrete method.

simplified snippet:
```scala
  trait Ops[F[_], A] extends UniversalTrait[F, A] with Serializable {}

  trait UniversalTrait[F[_], A] extends Any {
    type TypeClassType <: Apply[F]
    def typeClassInstance: TypeClassType
    def self: F[A]
    def ap[B, C](fa: F[B])(implicit ev$1: A <:< (B => C)): F[C] =
      typeClassInstance.ap[B, C](self.asInstanceOf[F[B => C]])(fa)
    def productR[B](fb: F[B]): F[B] = typeClassInstance.productR[A, B](self)(fb)
    def productL[B](fb: F[B]): F[A] = typeClassInstance.productL[A, B](self)(fb)
    @inline final def <*>[B, C](fa: F[B])(implicit ev$1: A <:< (B => C)): F[C] =
      typeClassInstance.<*>[B, C](self.asInstanceOf[F[B => C]])(fa)
    @inline final def *>[B](fb: F[B]): F[B] = typeClassInstance.*>[A, B](self)(fb)
    @inline final def <*[B](fb: F[B]): F[A] = typeClassInstance.<*[A, B](self)(fb)
    def ap2[B, C, D](fa: F[B], fb: F[C])(implicit ev$1: A <:< ((B, C) => D)): F[D] =
      typeClassInstance.ap2[B, C, D](self.asInstanceOf[F[(B, C) => D]])(fa, fb)
    def map2[B, C](fb: F[B])(f: (A, B) => C): F[C] = typeClassInstance.map2[A, B, C](self, fb)(f)
    def map2Eval[B, C](fb: Eval[F[B]])(f: (A, B) => C): Eval[F[C]] = typeClassInstance.map2Eval[A, B, C](self, fb)(f)
  }
```
then mima says:
```
...
[error]  * in current version, classes mixing cats.Apply$Ops need be recompiled to wire to the new static mixin forwarder method all super calls to method ap(java.lang.Object,scala.Predef#<:<)java.lang.Object
[error]    filter with: ProblemFilters.exclude[NewMixinForwarderProblem]("cats.Apply#Ops.ap")
[error]  * in current version, classes mixing cats.Apply$Ops need be recompiled to wire to the new static mixin forwarder method all super calls to method productR(java.lang.Object)java.lang.Object
[error]    filter with: ProblemFilters.exclude[NewMixinForwarderProblem]("cats.Apply#Ops.productR")
[error]  * in current version, classes mixing cats.Apply$Ops need be recompiled to wire to the new static mixin forwarder method all super calls to method productL(java.lang.Object)java.lang.Object
[error]    filter with: ProblemFilters.exclude[NewMixinForwarderProblem]("cats.Apply#Ops.productL")
...
```

Reply:
Ah yeah.. I remember I ran into it too some time ago... The thing is that if we had real implementations behind those concrete methods, then there's a trick to keep them.... But since they are just redirects to the type class implementations, then it is just not worth it.

Reply:
...on the bright side, I think we could try to benefit from it. I mean, if we cannot (and don't have to) simply copy the old `Ops` definition, then we could make a new one nicer and neater than the old one.

E.g. instead of having this bulky and hacky
```scala
def ap[B, C](fb: F[B])(implicit ev$1: A <:< (B => C), F: Apply[F]): F[C] =
    F.ap(fa.asInstanceOf[F[B => C]])(fb)
```
we could try extracting it into a separate value class, like this
```scala
final class ApplyFAB[F[_], A, B](private val fab: F[A => B]) extends AnyVal {
  def ap(fa: F[A])(implicit F: Apply[F]): F[B] = F.ap(fab)(fa) 
}
```
Since we already have a similar thing for some special cases like `F[Boolean]` down below, so we could try to make the syntax even better, wdyt?

Reply:
...potentially it can break source compatibility though, but I guess it is only for very special cases where someone calls `ap` and passes all the implicits explicitly...

Reply:
Yes, that's a good idea. We should take this opportunity to do some tidying.

Reply:
thank you for your support :)

I did some tidying for `Apply`, so could you take a look whether I'm on the right track?
Once `Apply` looks ok, I will make similar modifications for other data types.